requests-for-comments/2025-04-15-scripts-to-rule-them-all.md

```diff
@@ -0,0 +1,72 @@
+
+I'm not suggesting we do _exactly_ what they do, but a similar approach would be a big benefit. Not every repository will need every script. Some
+repositories will need scripts custom to the environment. That's all fine. The goal is to have a baseline set of scripts that we can use to get a
+project up and running in a development environment and a known location to look for those scripts.
+
+## Context
+
+For context, here's the list of scripts referenced in the GitHub blog post ([samples available in a repository](https://github.com/github/scripts-to-rule-them-all)):
+
+* `script/bootstrap` - Install dependencies (Ex. npm packages, brew packages, etc. Usually run once after cloning the repository)
+* `script/setup` - Setup the project. Seems like this could just be merged in bootstrap
+* `script/update` - Update project after a pull such as migrations, etc.
+* `script/server` - For web apps, starts the application. For client SDKs, might start an example site.
+* `script/test` - Run tests (Ex. `npm test`, `bundle exec rspec`, etc.). Also includes linting, formatting, etc.
+* `script/cibuild` - used by the CI system to run tests, build, etc.
+* `script/console` - is used to open a console for your application.
+
+## Design
+
+For `posthog/posthog`, we already have a set of scripts in the `bin` directory. One issue with using `bin` across the board is that some environments
+add `bin` to the `.gitignore` file by default because that's where they compile binaries to. However, we've already set this precedence, so it may make
+sense to stick to it. For example, [posthog/posthog-dotnet](https://github.com/PostHog/posthog-dotnet/tree/main/bin) already has a `bin` directory.
+
+* `bin/setup` - Install or upgrade dependencies (Ex. npm packages, brew packages, etc. Usually run once after cloning the repository).
+* `bin/update` - Updates dependencies after a pull. This could simply call `bin/setup`.
+* `bin/build` - Build the project, for projects that are compiled such as C#, Java, etc.
+* `bin/start` - Start the project. For SDKs, this might start an example server.
+* `bin/test` - Run tests (Ex. `npm test`, `bundle exec rspec`, etc.). Also includes linting, formatting, etc.
+* `bin/fmt` - Optional: Format/lint code. This can be called by `test`.
+
+## Alternative Ideas Considered
+
+I've noticed a lot of our repos have a `make` file. This is especially common for mobile development. An alternate proposal would be to standardize on `make` rather than these `bin` scripts.
+
+My recommendation is to still standardize on `bin` scripts so that it's consistent across the board. Some of the bin scripts are pretty complicated and it'll be hard to port to `make`.
+If some repos have `make` files, that's fine. The bin scripts can call the `make` commands.
+
+## Feedback Requested
+
+* What other scripts do we need?
+* What should the scripts be named?
+* Should we use `bin` or something else?
```

Comment:
I'd rather use just `scripts`, `bin` is used by some other languages/frameworks (eg a dart project will have its executable and other entry points in the bin folder) as part of the folder structure and our scripts will be among many different files, a bit confusing

Reply:
It's too bad Swift doesn't have union types. In practice, a feature flag result will only be `null`, `bool`, or a `string`. If `getFlagValue` returned something other than one of those types, that would be unexpected. In that case, I think returning `false` is appropriate OR we should throw an exception. What do you think?

Reply:
I'll let you and @rafaeelaudibert debate `bin` vs `scripts`. 😆

Reply:
I vouched for `bin` because we're already using and I don't wanna hurt our muscle memory.

It is true `bin` is used in stuff like Ruby/Rails as well to include default scripts, but isn't this just the same as we already have here? The only difference is that they might have some extra files inside `bin` that don't exist in Django/JS land, for example

If that's not the case, and they're more than just scripts, I'm fine changing to scripts. it will hurt in the short-term, but it doesnt really matter in the long-term

Reply:
Some languages/frameworks have `bin` as part of .gitignore btw, you'd need to remove from gitignore and add all the other files that should not be checked under the source control.
no strong opinions here, I just think that `scripts` would be cleaner and avoid this known pitfall.
for muscle memory, you could either set up your own alias or even have a symbolic file linked to the correct one so if you do eg ./lint in the root folder, it just works it does not matter if its within the bin or scripts really